src/content/en/updates/posts/2016/07/infinite-scroller.md

```diff
@@ -0,0 +1,246 @@
+
+<img src="/web/updates/images/2016/07/infinite-scroller/tombstone.png" alt="Such tomb. Very stone. Wow.">
+
+An interesting challenge here is that real items can have a bigger height than the tombstone item because of differing amounts of text per item or an attached image. To resolve this, we will adjust the current scroll position every time data comes in and a tombstone is being replaced above the viewport, *anchoring* the scroll position to an element rather than a pixel value. This concept is called scroll anchoring.
+
+#Scroll Anchoring
+Our scroll anchoring will be invoked both when tombstones are being replaced as well as when the window gets resized (which also happens when the devices is being flipped!). We will have to figure out what the top-most visible element in the viewport is. As that element could only be partially visible, we will also store the offset from the top of the element where the viewport begins.
+
+<img src="/web/updates/images/2016/07/infinite-scroller/anchoring.png">
+
+After things in the runway have changed, we can now restore a situation that feels almost identical to the user. Win! Except a resized window means that each items has potentially changed its height, so how do we know how far down the anchored content should be placed? We don’t! To find out we would have to layout every element above the anchored item and add up all of their heights; this could cause a significant pause after a resize, and we don’t want that. Instead, we resort to assuming that every item above is the same size as a tombstone and adjust our scroll position accordingly. As elements are scrolled into the runway, we adjust or scroll position, effectively deferring the layout work to when it is actually needed.
+
+#Layout
+I have skipped over an important detail: Layout. Each recycling of a DOM element would normally relayout the entire runway which would bring us well below our target of 60 frames per second. To avoid this, we are taking the burden of layout onto ourselves and use absolutely positioned elements with transforms. This way we can pretend that all the elements further up the runway are still taking up space when in actuality there is only empty space. Since we are doing layout ourselves, we can cache the positions each item ends up and we can immediately load the correct element from cache when the user scrolls backwards.
```

Comment:
'we can cache the positions each item ends up' > 'we can cache the positions where each item ends up'